### PR TITLE
Implement queue Capability

### LINKED ISSUE #117: MaxResource of Queue
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**Description**:

Support MaxResource of queue by admission controller & queue controller. 

xref https://github.com/kubernetes-sigs/kube-batch/issues/452

- [ ] Update `Queue` API (https://github.com/volcano-sh/volcano/issues/131)
- [ ] Added QueueController (https://github.com/volcano-sh/volcano/issues/16)
- [ ] ~~Added Queue admission controller~~

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Partially fix https://github.com/volcano-sh/volcano/issues/117

**Special notes for your reviewer**:

**Release note**:

```release-note
donot allow podgroup enqueue when queue capability reached
```

### CODE REVIEW

File: pkg/scheduler/actions/enqueue/enqueue.go [hunk 1]
```diff
@@ -100,14 +100,14 @@ func (enqueue *enqueueAction) Execute(ssn *framework.Session) {
...
 			inqueue = true
 		} else {
-			if job.PodGroup.Spec.MinResources == nil {
-				inqueue = true
-			} else {
-				pgResource := api.NewResource(*job.PodGroup.Spec.MinResources)
-
+			pgResource := api.NewResource(*job.PodGroup.Spec.MinResources)
+			// The queue resource quota limit has not reached
+			if pgResource.Clone().Add(ssn.Queues[queue.UID].Allocated).LessEqual(api.NewResource(queue.Queue.Spec.Capability)) {
```
Comment (k82cn) [inline pkg/scheduler/actions/enqueue/enqueue.go]: we should add an `Enqueue` callback to ask plugin whehter it's ready to enqueue, refer to JobReady callback.

Comment (hzxuzhonghu) [reply pkg/scheduler/actions/enqueue/enqueue.go]: Will have a look

Comment (hzxuzhonghu) [reply pkg/scheduler/actions/enqueue/enqueue.go]: done

File: pkg/scheduler/actions/enqueue/enqueue.go [hunk 2]
```diff
@@ -100,18 +100,14 @@ func (enqueue *enqueueAction) Execute(ssn *framework.Session) {
...
+			if ssn.JobEnqueueable(job) && pgResource.LessEqual(nodesIdleRes) {
+				nodesIdleRes.Sub(pgResource)
 				inqueue = true
-			} else {
-				pgResource := api.NewResource(*job.PodGroup.Spec.MinResources)
-
-				if pgResource.LessEqual(nodesIdleRes) {
-					nodesIdleRes.Sub(pgResource)
-					inqueue = true
-				}
```
Comment (k82cn) [inline pkg/scheduler/actions/enqueue/enqueue.go]: If `inqueue`, why push queue back?

File: pkg/scheduler/plugins/proportion/proportion.go [hunk 3]
```diff
@@ -198,6 +198,19 @@ func (pp *proportionPlugin) OnSessionOpen(ssn *framework.Session) {
...
 	})
 
+	ssn.AddJobEnqueueableFn(pp.Name(), func(obj interface{}) bool {
+		job := obj.(*api.JobInfo)
+		queueID := job.Queue
+		attr := pp.queueOpts[queueID]
+		queue := ssn.Queues[queueID]
+		pgResource := api.NewResource(*job.PodGroup.Spec.MinResources)
+		// The queue resource quota limit has not reached
+		if pgResource.Clone().Add(attr.allocated).LessEqual(api.NewResource(queue.Queue.Spec.Capability)) {
```
Comment (k82cn) [inline pkg/scheduler/plugins/proportion/proportion.go:208]: We may also consider maxPodGroup number as one of resource, we can open another PR for it.

**Comment by k82cn** [comment]
LGTM overall, please ping me when CI is happy with that.

File: pkg/scheduler/plugins/proportion/proportion.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pkg/scheduler/plugins/proportion/proportion.go:208]: ok

**Review by k82cn — APPROVED** [review]

**Pull request merged**